ruby/ext/google/protobuf_c/encode_decode.c

```diff
@@ -645,21 +651,19 @@ VALUE Message_decode(VALUE klass, VALUE data) {
 
   const upb_pbdecodermethod* method = msgdef_decodermethod(desc);
   const upb_handlers* h = upb_pbdecodermethod_desthandlers(method);
-  upb_pbdecoder decoder;
-  upb_sink sink;
-  upb_status status = UPB_STATUS_INIT;
+  upb_env e;
+  upb_env_init(&e);
```

Comment:
If the overhead of `malloc()` is noticeable here (and it's possible that it's not, at all; I'm curious to know), there are two options for avoiding this overhead:

1. You can give the upb_env a custom allocator that is seeded with some stack memory. The code would be:

    char stack_mem[UPB_PB_DECODER_SIZE];
    upb_seededalloc alloc;
    upb_seededalloc_init(&alloc, stack_mem, sizeof(stack_mem));
    upb_env e;
    upb_env_init(&e);
    upb_env_setallocfunc(&e, upb_seededalloc_getallocfunc(&alloc), &alloc);

    upb_env_uninit(&e);
    upb_seededalloc_uninit(&alloc);

    Definitely more verbose and less convenient, so I only recommend this if there is a measurable difference.

2. Reuse a upb_env across encodes. If you have any convenient place to store a upb_env that has already been built (all the pipeline elements created and plumbed together), then you can just reuse the same one each time.

But I suspect the perf difference here will not be too noticeable, given how much overhead Ruby is imposing already.